### PR TITLE
verify check block and log block

### PR DESCRIPTION
**Pull request by FelixFan1992**

### CODE REVIEW

**Comment by FelixFan1992** [comment]
this PR aims to verify check block and log block:
if check block is too old or check block/hash combo does not exist anymore, return error
if log block is not populated, make a RPC call to verify tx hash still exists
otherwise, verify log block/hash combo exists on chain via block subscriber

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go [hunk 1]
```diff
@@ -926,3 +937,75 @@ func (r *EvmRegistry) getBlockHash(blockNumber *big.Int) (common.Hash, error) {
...
+		WorkID:        p.WorkID,
+		FastGasWei:    big.NewInt(0),
+		LinkNative:    big.NewInt(0),
+	}
+}
+
+// verifyCheckBlock checks that the check block and hash are valid
+func (r *EvmRegistry) verifyCheckBlock(checkBlock, upkeepId *big.Int, checkHash common.Hash, p ocr2keepers.UpkeepPayload, i int, results []ocr2keepers.CheckResult) bool {
+	// verify check block number is not too old
+	if r.bs.latestBlock-checkBlock.Int64() > validCheckBlockRange {
```
Comment (amirylm) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:964]: accessing `r.bs.latestBlock` like that can lead to a race..

using `atomic.Int64` for `latestBlock` sounds like the best solution,
but you can also wrap it with a function, e.g. `getLatestBlock` and use a read lock to protect.

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go [hunk 2]
```diff
@@ -926,3 +937,75 @@ func (r *EvmRegistry) getBlockHash(blockNumber *big.Int) (common.Hash, error) {
...
+		r.lggr.Warnf("latest block is %d, check block number %s is too old for upkeepId %s", r.bs.latestBlock, checkBlock, upkeepId)
+		results[i] = r.getCheckResult(p, UPKEEP_FAILURE_REASON_CHECK_BLOCK_TOO_OLD)
+		return false
+	}
+	// verify check block number and hash are valid
+	h, ok := r.bs.blocks[checkBlock.Int64()]
```
Comment (amirylm) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:970]: accessing the map of block subscriber could lead to a data race

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go (hunk 1 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:964]: done

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go (hunk 2 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/registry.go]: done

**Pull request merged**